[serve] Correct error detection and add more log messages

jugglinmike opened the pull request:
A more advanced solution might involve attempting to re-start the failed
subprocess(es). I think we'd need to understand the failure conditions before
committing to something like that, though. Otherwise, we'll risk enabling a
"continual restart" state if the problem can't be resolved by creating a new
sub-process.

I wrote the new test to a dedicated file because as an integration-level test,
it's quite a bit slower than the existing tests. It's also a little messy
because it stubs out a module method. I'm open to suggestions about better ways
to test this!

jugglinmike committed:
[serve] Quit on sub-process failure

The correctness of the WPT server is dependent on the availability of
all server sub-processes. If any sub-process dies, the server is no
longer capable of performing its role. Allow the process to exit in
this case.

Introduce a test for this behavior.

jugglinmike committed:
[serve] Log state of subprocesses during shutdown

jugglinmike commented:
The new test interacts with the existing tests because the server implements logging as a singleton:

>     Exception in thread Thread-1:
>     Traceback (most recent call last):
>       File "/usr/lib/python2.7/threading.py", line 801, in __bootstrap_inner
>         self.run()
>       File "/usr/lib/python2.7/threading.py", line 754, in run
>         self.__target(*self.__args, **self.__kwargs)
>       File "/home/mike/projects/bocoup/google-wpt/web-platform-tests/tools/serve/test_functional.py", line 49, in target
>         serve.run(config_path=tempfile_name)
>       File "/home/mike/projects/bocoup/google-wpt/web-platform-tests/tools/serve/serve.py", line 809, in run
>         set_logger(logger)
>       File "/home/mike/projects/bocoup/google-wpt/web-platform-tests/tools/wptserve/wptserve/logger.py", line 23, in set_logger
>         raise Exception("Logger must be set at most once")
>     Exception: Logger must be set at most once

I'm out of time for the day, but I'll try to find a workaround by Monday.

foolip commented on tools/serve/serve.py:
```diff
@@ -378,15 +378,17 @@ def build_routes(aliases):
 
 class ServerProc(object):
-    def __init__(self):
+    def __init__(self, name=None):
```
If any other review wonders, yes there is another case `wrapper = ServerProc()` where `name` isn't given, and that seems fine.

foolip commented on tools/serve/serve.py:
```diff
@@ -827,12 +829,15 @@ def run(**kwargs):
                 logger.info("Shutting down")
 
+        for item in iter_procs(servers):
+            logger.info("Status of %s:\t%s" % (item.name, "running" if item.is_alive() else "not running"))
```
If `item.name` is `None` here the message won't be great. Is that impossible? If not, maybe just require `name` after all?

jugglinmike replied:
Thanks for pointing this out. Including the port number in the name is a litle redundant since that's already defined via an instance property. Because that property *is* available on all instances, we can name the process with something slightly useful even when additional information (i.e. the scheme) is not specified.

foolip commented on tools/serve/serve.py:
```diff
@@ -827,12 +829,15 @@ def run(**kwargs):
             except KeyboardInterrupt:
                 logger.info("Shutting down")
 
+        for item in iter_procs(servers):
```
A sample of the output I see:
```
^CINFO:web-platform-tests:Shutting down
INFO:web-platform-tests:Status of ws on port 61486:	running
INFO:web-platform-tests:Status of wss on port 61487:	running
INFO:web-platform-tests:Status of http on port 8000:	not running
INFO:web-platform-tests:Status of http on port 61488:	not running
INFO:web-platform-tests:Status of https on port 8443:	not running
```

As a user of `./wpt serve` manually, I'm not sure what to do with this on shutdown, if it's cause for concern or not. Do you think these make more sense as debug messages?

Alternatively, if the reason we shut down is that some subprocess has died, can we log that fact clearly, and just in that case?

jugglinmike replied:
I prefer the second alternative because this information does seem relevant to the typical user of `wpt serve`, just not so much in the "Ctrl-C" case. Here's what it looks like now (printed only in response to subprocess exit):

>     INFO:web-platform-tests:2 subprocesses exited:
>     INFO:web-platform-tests:Status of ws on port 44283:     running
>     INFO:web-platform-tests:Status of wss on port 44819:    running
>     INFO:web-platform-tests:Status of http on port 8000:    not running
>     INFO:web-platform-tests:Status of http on port 40161:   running
>     INFO:web-platform-tests:Status of https on port 8443:   not running

foolip reviewed (commented):
I've reviewed without the final "DEBUG" commit, assuming nothing in that commit should stay.

foolip commented on tools/serve/serve.py:
```diff
@@ -827,9 +829,17 @@ def run(**kwargs):
+                logger.info("%s %s exited:" % (len(exited), subject))
+
+                for item in iter_procs(servers):
+                    logger.info("Status of %s:\t%s" % (item.name, "running" if item.is_alive() else "not running"))
```
Ah, when removed from the ctrl+c case, I agree this is useful info. Thanks!

jugglinmike commented:
Regarding the trouble with the logger: I initially tried to resolve this by running the test in a subprocess, but [the particulars of multiprocessing in Python](https://kaushikghose.wordpress.com/2016/08/26/python-global-state-multiprocessing-and-other-ways-to-hang-yourself/) mean that "use a subprocess" is not a silver bullet.

Instead, I'm using `reload` to forcibly reset the logging module's internal state prior to testing. Just like overriding `ServerProc`, this seems like an unhealthy coupling between the tests and the implementation, but I can't think of a more hygienic solution.

This branch accrued a number of debugging commits because working this out took a bit of trial-and-error. My apologies to the folks who received notifications for all that. I've rebased and squashed everything back down (including the changes @foolip recommended). [The original version of this branch is available here.](https://github.com/web-platform-tests/wpt/compare/master...bocoup:serve-shutdown-prebase) This branch should be ready to merge.
